src/intersection-observer-polyfill.js

```diff
@@ -33,6 +34,133 @@ import {layoutRectLtwh, rectIntersection, moveLayoutRect} from './layout-rect';
  */
 export let DOMRect;
 
+export const AMP_DEFAULT_THRESHOLD =
+    [0, 0.05, 0.1, 0.15, 0.2, 0.25, 0.3, 0.35, 0.4,
+    0.45, 0.5, 0.55, 0.6, 0.65, 0.7, 0.75, 0.8, 0.85, 0.9, 0.95, 1];
+
+/**
+ * A function to get the element's current IntersectionObserverEntry
+ * regardless of the intersetion ratio. Only available when element is not
+ * nested in a container iframe.
+ * @param {!./layout-rect.LayoutRectDef} element element's rect
+ * @param {?./layout-rect.LayoutRectDef} owner element's owner rect
+ * @param {!./layout-rect.LayoutRectDef} hostViewport hostViewport's rect
+ */
+export function getIntersectionChangeEntry(
+    element, owner, hostViewport) {
+  const intersection = calculateIntersectionRect(
+        element, owner, hostViewport);
+  const ratio = intersectionRatio(intersection, element);
+  return calculateChangeEntry(
+      element, hostViewport, intersection, ratio);
+}
+
+/**
+ * A class to help amp-iframe and amp-ad nested iframe listen to intersection
+ * change.
+ */
+export class IntersectionObserverApi {
+  /**
+   * @param {!AMP.BaseElement} baseElement
+   * @param {!Element} iframe
+   * @param {boolean=} opt_is3p
+   */
+  constructor(baseElement, iframe, opt_is3p) {
+    /** @private @const {!AMP.BaseElement} */
+    this.baseElement_ = baseElement;
+
+    /** @private {?IntersectionObserverPolyfill} */
+    this.intersectionObserver_ = null;
+
+    /** @private {!boolean} */
+    this.shouldObserve_ = false;
+
+    /** @private {!boolean} */
+    this.isInViewport_ = false;
+
+    /** @private {?function()} */
+    this.unlistenOnDestroy_ = null;
+
+    /** @private @const {!./service/viewport-impl.Viewport} */
+    this.viewport_ = baseElement.getViewport();
+
+    /** @private {?SubscriptionApi} */
+    this.subscriptionApi_ = new SubscriptionApi(
+        iframe, 'send-intersections', opt_is3p || false, () => {
+          this.startSendingIntersection_();
+        });
+
+    this.intersectionObserver_ = new IntersectionObserverPolyfill(change => {
+      this.subscriptionApi_.send('intersection', {changes: [change]});
+    }, {threshold: AMP_DEFAULT_THRESHOLD});
+  }
+
+  /**
+   * Function to start listening to viewport event. and observer intersection
+   * change on the element.
+   */
+  startSendingIntersection_() {
+    this.shouldObserve_ = true;
+    this.intersectionObserver_.observe(this.baseElement_.element);
+    this.baseElement_.getVsync().measure(() => {
+      if (this.baseElement_.isInViewport()) {
+        this.isInViewport_ = true;
+        this.intersectionObserver_.tick(this.viewport_.getRect());
+      }
+    });
+
+    const unlistenViewportScroll = this.viewport_.onScroll(() => {
+      if (!this.isInViewport_) {
```

Comment:
The `#onScroll` callback can just be a bound `#onLayoutMeasure` (as we currently call it). It used to be called `#fire`.